modules/core/src/main/java/org/apache/ignite/internal/util/VersatileReadWriteLock.java

```diff
@@ -364,16 +364,31 @@ private void notifyWriteLockSolicitors() {
         for (Iterator<CompletableFuture<Void>> iterator = writeLockSolicitors.iterator(); iterator.hasNext(); ) {
             CompletableFuture<Void> future = iterator.next();
 
-            if (!tryWriteLock()) {
+            if (tryWriteLock()) {
+                // We were able to acquire the write lock. It might be that lock intent corresponding to this future has already been
+                // handled by another thread (and the lock has already been released), so we need to check this; if this is true,
+                // we will need to release the lock.
+                // We can use non-atomic pattern 'check whether future is done, and if not, finalize acquisition and complete the future'
+                // because this is done in the critical section (under protection 'holding the write lock' invariant).
+
+                // Removing as soon as possible to avoid an infinite recursion in the #writeUnlock() call that follows.
+                iterator.remove();
+
+                if (!future.isDone()) {
+                    // First finalize the aqcuisition.
+                    decrementPendingWriteLocks();
+
+                    asyncContinuationExecutor.execute(() -> future.complete(null));
+                } else {
+                    // The one who added this future has already taken the write lock for us (and it has already been released
+                    // [as we could take it again]; as they have completed the future, this is us who needs to unlock the excess.
+                    // They also decremented pending write locks, so we don't need to do it again.
+                    writeUnlock();
+                }
+            } else {
```

Comment:
Can we invert the `if` condition and move it to the top as the fast path? It will make the code more readable